packages/flutter/lib/src/scheduler/binding.dart

```diff
@@ -199,18 +199,55 @@ mixin SchedulerBinding on BindingBase, ServicesBinding {
 
     if (!kReleaseMode) {
       int frameNumber = 0;
-
-      // use frameTimings. https://github.com/flutter/flutter/issues/38838
-      // ignore: deprecated_member_use
-      window.onReportTimings = (List<FrameTiming> timings) {
+      addTimingsCallback((List<FrameTiming> timings) {
         for (FrameTiming frameTiming in timings) {
           frameNumber += 1;
           _profileFramePostEvent(frameNumber, frameTiming);
         }
-      };
+      });
+    }
+  }
+
+  /// Add a [TimingsCallback] that receives [FrameTiming] sent from the engine.
+  ///
+  /// This can be used, for example, to monitor the performance in release mode,
+  /// or to get a signal when the first frame is rasterized.
+  ///
+  /// This is preferred over using [Window.onReportTimings] directly because
+  /// [addTimingsCallback] allows multiple callbacks.
+  ///
+  /// If the same callback is added twice, it will be executed twice.
+  void addTimingsCallback(TimingsCallback callback) {
+    // TODO(liyuqian): once this is merged, modify the doc of
+    //  [Window.onReportTimings] inside the engine repo to recommend using this
+    // API instead of using [Window.onReportTimings] directly.
+    _timingsCallbacks.add(callback);
+    if (_timingsCallbacks.length == 1) {
+      assert(window.onReportTimings == null);
+      window.onReportTimings = _executeTimingsCallbacks;
     }
+    assert(window.onReportTimings == _executeTimingsCallbacks);
   }
 
+  /// Removes a callback that was earlier added by [addTimingsCallback].
+  void removeTimingsCallback(TimingsCallback callback) {
+    assert(_timingsCallbacks.contains(callback));
+    _timingsCallbacks.remove(callback);
+    if (_timingsCallbacks.isEmpty) {
+      window.onReportTimings = null;
+    }
+  }
+
+  void _executeTimingsCallbacks(List<FrameTiming> timings) {
+    final List<TimingsCallback> clonedCallbacks =
+    List<TimingsCallback>.from(_timingsCallbacks);
+    for (TimingsCallback callback in clonedCallbacks) {
+      callback(timings);
```

Comment:
In other places where we have similar code we check here if the callback is still in clonedCallbacks, for example: https://github.com/flutter/flutter/blob/b31ca1aaca0b9ef70ec478a74b3b690e4d632616/packages/flutter/lib/src/foundation/change_notifier.dart#L205-L206

That way if your list is [callbackA, callbackB] and callbackA removes callbackB, callbackB is not called again. This would probably also make sense here?

Also from that link: Should we catch exceptions here that happen while executing callback?

Reply:
Done. Also added a test for the error handling.